### ISSUE TITLE
A required argument overriding an optional argument in the `ParamSpec` is incorrectly marked as optional when this argument is introduced by referencing `__init__`.

### ISSUE THREAD

**Issue by cainmagi** [microsoft/pyright#9239]
This issue is a follow-up of the previous issue, see the following talk:

> Yes, that's a different issue from the one I fixed and is only partly related to the original bug you posted in this issue. Please post a new issue so it doesn't get lost. 

 _Originally posted by @erictraut in [#8964](https://github.com/microsoft/pyright/issues/8964#issuecomment-2379605070)_

## Describe the bug

A function using a `ParamSpec` can define a required argument that has the same name of an optional argument in the `ParamSpec`. In this case, this function supposes to raise an issue if the argument is not provided.

The above logic works correctly for a normal function. However, if the function is a method from a mixin class, `pyright` will not report this issue.

## Code or Screenshots

Pyright Demo: (pyright only reports one issue)

https://pyright-play.net/?pyrightVersion=1.1.385&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAMLkA2D5ARg8TQAqXkQDKCYgGMaIYgDdijAPrxBAKHmcoAXijcQvAcIAUAIk56AlPIKrCJCiH0Fji%2BUOYBnJ1ACySAB6oAXPKgBUAACjuQuEMQwABZgACb%2BgbHEwFDA5EIwuHA6jk4%2BdIzMbMQA2pw0BAC6NLHkMOT5TjDUUABUlGh56gB0HU40ra0A1gDuffmc3aN9RlAAtAB8hH6Bq1AIIKgwOmIbOrX1RqLEezPHe9MgnUazUADEUADkMMSeMI9QOrMA3gC%2BCWsoGIYABXEAoKC5HTtK79NrDMawkyKUIuKAAOVe2w83hQRhWiWSUGk0lQSBgJJ0TmIDGANTqDUwKHwagAjLNFhiwChiATAQgwk57EkUjy3gAmfYM-Jbczs%2BZLdHc3kA9aC4VEtIZLKS3L5ehMVjsMoVapQA6MpotGGdCa9WEDBHjHqXa4Krk8vlqlz2MSSGRyYg6TFvbpazIgOAmcSMcwhmBh9IR7K3B58GIghixIHkJDUqDkCF5pwg4hQKLEMTdQhRPNQBioMuxMDEVwoMD4Ea4IZMoEgiEwLDEbryP1SBiyEg6GMMZHhnU6MUwcWpqAAJVz%2BcLmBcpYLrleggydSQ3JHQA

Mypy Demo: (mypy can report both issues correctly)

https://mypy-play.net/?mypy=1.12.0&python=3.13&gist=7b7e3a1c0f9836912d6f751f49948e3f

Codes:

```python
from typing import TypeVar, Callable, ParamSpec, reveal_type

P = ParamSpec("P")
T = TypeVar("T")


class Mixin:
    @classmethod
    def factory(cls: Callable[P, T], data: str, *args: P.args, **kwargs: P.kwargs) -> T:
        print(repr(data), repr(args), repr(kwargs))  # 'text' () {}
        return cls(*args, **kwargs)


class Next(Mixin):
    def __init__(self, data: int = 1) -> None:
        pass


def next2(data: int = 1) -> None:
    pass


def factory2(cls: Callable[P, T], data: str, *args: P.args, **kwargs: P.kwargs) -> None:
    pass


reveal_type(Next.factory)
val = Next.factory()  # Should raise an issue here. This line does not work in run time.
reveal_type(val)

factory2(next2)  # Raise an issue as expectation.
```

## Another finding

Change the usage of `Next.factory()` by this:

```python
val = Next.factory(1)  # the correct type of this input argument should be `str`
```

pyright still does not report it as an issue. However, mypy can detect that

```
main.py:28: error: Argument 1 to "factory" of "Mixin" has incompatible type "int"; expected "str"  [arg-type]
```

## VS Code extension or command-line

* `pyright`: 1.1.382 (the playground version is 1.1.385)
* `pylance`: 2024.10.1
* `python`: 3.12.2.final.0
* `VSCode`: 1.94.2
* `OS`: Windows x64 10.0.22631

**Comment by erictraut** [comment]
I think pyright's behavior here is correct. When `factory` is bound to `Next`, the resulting type is partially specialized, and `P` and `T` receive values due to the binding.

The exact details of binding and partial specialization are not clearly spelled out in the typing spec currently, but I think pyright's behavior here is arguably more correct and internally consistent than mypy's. 

In any case, pyright is working as intended here, so I don't consider this a bug.

**Comment by cainmagi** [comment]
I do not understand why this behavior is correct. Running the above codes in run time will raise the following error.

```
Traceback (most recent call last):
  File "/workdir/test.py", line 28, in <module>
    val = Next.factory()  # Should raise an issue here. This line does not work in run time.
          ^^^^^^^^^^^^^^
TypeError: Mixin.factory() missing 1 required positional argument: 'data'
```

I think `mypy` is correct because its behavior is apparently consistent with the run time behavior, while `pyright` does not.

Do you mean that `pyright` does not raise the error because the user's codes do not correctly specify the type hints? If so, is it more appropriate to raise an issue to let users know that the type hint definition is not reasonable?

**Comment by erictraut** [comment]
Apologies, my initial analysis was incorrect.

I didn't notice that both the `factory` and `__init__` methods have parameters named `data`. That's the source of the problem here. The partially-specialized type of `Next.factory` is `(data: str, data: int=1) -> Next`. 

I'm reopening the issue, and I'll work on a fix.

**Comment by cainmagi** [comment]
Thank you! I thinks this is the last unexpected behavior that I can find.

**Comment by erictraut** [comment]
For clarity, this issue is not related to mixin classes. Here's a minimal repro:

```python
from typing import Callable

class A:
    def __init__(self, data: int = 1) -> None:
        pass

    @classmethod
    def method[**P, T](cls: Callable[P, T], data: str, *args: P.args, **kwargs: P.kwargs) -> T: ...

val = A.method()
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.386
